Add C++ sample robot code

prateekma opened the pull request:
Pretty much translated the Java equivalent to C++ to the best of my abilities. Haven't tested this on a robot yet, but will do so when I get access to one of our robots.

Let me know if there are any errors that I should fix.

virtuald commented:
Since we distribute this to high school students, I'd prefer that someone at least verifies that it compiles before we merge it.

virtuald commented on robot-java/src/main/java/dc/Robot.java:
```diff
@@ -157,7 +154,7 @@ public void autonomousPeriodic() {
 
 		// command motors to do things
-drive.tankDrive(autospeed, autospeed, false);
+		drive.tankDrive(autospeed, autospeed, false);
```
Why did this file change? It's not relevant to your PR.

prateekma replied:
My bad, I probably auto-formatted it by accident; I'll revert this change.

auscompgeek commented:
> Since we distribute this to high school students, I'd prefer that someone at least verifies that it compiles before we merge it.

It seems like we should have these sorts of checks in CI.

prateekma commented:
I've verified that both projects compile by running a `./gradlew build`. I will test out the actual code on a robot tonight if I get the chance.

prateekma commented:
I ran the `robot-cpp-talonsrx` on our robot tonight and found one bug that I fixed and committed. It involved the follower TalonSRX's not being added to SpeedControllerGroup. I thought this wasn't needed because they would follow the leader, but apparently not.

After that change was made, the program worked as intended with `data_logger.py`. I was unable to test out the version with Sparks since we don't have a robot that has Sparks on it. Fortunately, I added the follower motors to the speed controller group in that version of the program.

auscompgeek commented:
> It involved the follower TalonSRX's not being added to SpeedControllerGroup. I thought this wasn't needed because they would follow the leader, but apparently not.

I think you've debugged this the wrong way. In the same commit, you kept the object around, whereas originally you had an object that got destructed after RobotInit. Following a Talon SRX in PercentOutput mode should work without any problems.

auscompgeek commented:
Looking at the Java and Python versions, it's quite likely they also suffer from this bug, assuming the Java and Python bindings actually call the CCI destructor as well. (I suspect the Python one doesn't currently, which itself is a bug.)

prateekma commented:
Hmm, looking back in hindsight, you're right. The follower objects were destructed after RobotInit. This probably caused the bug I described.

I was a little confused because we've done similar things in Kotlin (JVM so basically the same as Java) i.e. declaring follower objects and setting them to follow the master motor in a configuration method. In this case, even when the method ends, the follower motor still continues to follow the motor, even though the master is the only object being kept around.

auscompgeek commented:
It's very much possible that the Phoenix Java library doesn't call the underlying destructor. I haven't taken a look at the Java bits this year yet (mostly on account of this year's stuff not being on GitHub until recently).

Oblarg commented:
Could you update this to reflect the new repo structure and include versions for arm characterization?

prateekma commented:
I will test the `robot-cpp-talonsrx` arm characterization on our robot soon (hopefully next week) to make sure that it works; but I simply adapted the `robot-cpp-talonsrx` drive characterization to reflect the changes needed to make it work with arm characterization.

auscompgeek commented on .gitignore:
```diff
@@ -7,7 +7,7 @@ bin
+**/.gradle
+**/.vscode
+**/.wpilib
+**/build
```
I think if we leave off the `**/` it'll work as expected too.

prateekma replied:
Hmm, without the second asterisk, Git resolves those folders as untracked. With the double asterisks, it ignores them.

This screenshot was with a single asterisk: http://prntscr.com/oaloid

auscompgeek replied:
That's why I'm saying leave off the globs entirely.

prateekma replied:
Ah yes, that does work. Fixed in 4752beb

auscompgeek commented on README.md:
```diff
@@ -13,8 +13,12 @@ Included Characterization Tools
+* [Python](robot-python)
+* [Python TalonSRX](robot-python-talonsrx)
+* [C++](robot-cpp)
+* [C++ TalonSRX](robot-cpp-talonsrx)
```
I think you want to change this only under the subdirectory readmes.

prateekma replied:
Whoops, my bad. Fixed this in latest commit.

auscompgeek commented:
Would it be better to have the number array as a local variable in AutonomousPeriodic? It's only used there and would avoid the length getting out of sync if anyone decides to send more data.

prateekma commented:
I can make the number array a static variable in AutonomousPeriodic. This will make it local, but it won't need to be constructed and destroyed every 20 milliseconds.

auscompgeek commented:
> This will make it local, but it won't need to be constructed and destroyed every 20 milliseconds.

A trivial allocation on the stack has no runtime cost.

If we have an actual local variable, the array can be initialised on declaration, and I believe we can omit the size in that case.

prateekma commented:
Tested the C++ code on our robot today and everything works as expected

virtuald commented:
Thank you for your contribution! Sorry for the delay.
